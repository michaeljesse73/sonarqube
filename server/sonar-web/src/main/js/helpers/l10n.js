/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
/* @flow */
import moment from 'moment';
import { getJSON } from './request';

let messages = {};

export function translate(...keys: string[]) {
  const messageKey = keys.join('.');
  return messages[messageKey] || messageKey;
}

export function translateWithParameters(messageKey: string, ...parameters: Array<string | number>) {
  const message = messages[messageKey];
  if (message) {
    return parameters
      .map(parameter => String(parameter))
      .reduce((acc, parameter, index) => acc.replace(`{${index}}`, parameter), message);
  } else {
    return `${messageKey}.${parameters.join('.')}`;
  }
}

export function hasMessage(...keys: string[]) {
  const messageKey = keys.join('.');
  return messages[messageKey] != null;
}

export function configureMoment(language?: string) {
  moment.locale(language || getPreferredLanguage());
}

function getPreferredLanguage() {
  return window.navigator.languages ? window.navigator.languages[0] : window.navigator.language;
}

function checkCachedBundle() {
  const cached = localStorage.getItem('l10n.bundle');

  if (!cached) {
    return false;
  }

  try {
    const parsed = JSON.parse(cached);
    return parsed != null && typeof parsed === 'object';
  } catch (e) {
    return false;
  }
}

function getL10nBundle(params) {
  const url = '/api/l10n/index';
  return getJSON(url, params);
}

export function requestMessages() {
  const browserLocale = getPreferredLanguage();
  const cachedLocale = localStorage.getItem('l10n.locale');
  const params = {};

  if (browserLocale) {
    params.locale = browserLocale;
  }

  if (browserLocale.startsWith(cachedLocale)) {
    const bundleTimestamp = localStorage.getItem('l10n.timestamp');
    if (bundleTimestamp !== null && checkCachedBundle()) {
      params.ts = bundleTimestamp;
    }
  }

  return getL10nBundle(params).then(
    ({ effectiveLocale, messages }) => {
      try {
        const currentTimestamp = moment().format('YYYY-MM-DDTHH:mm:ssZZ');
        localStorage.setItem('l10n.timestamp', currentTimestamp);
        localStorage.setItem('l10n.locale', effectiveLocale);
        localStorage.setItem('l10n.bundle', JSON.stringify(messages));
      } catch (e) {
        // do nothing
      }
      configureMoment(effectiveLocale);
      resetBundle(messages);
    },
    ({ response }) => {
      if (response && response.status === 304) {
        configureMoment(cachedLocale || browserLocale);
        resetBundle(JSON.parse(localStorage.getItem('l10n.bundle') || '{}'));
      } else {
        throw new Error('Unexpected status code: ' + response.status);
      }
    }
  );
}

export function resetBundle(bundle: Object) {
  messages = bundle;
}

export function installGlobal() {
  window.t = translate;
  window.tp = translateWithParameters;
  window.requestMessages = requestMessages;
}

export function getLocalizedDashboardName(baseName: string) {
  const l10nKey = `dashboard.${baseName}.name`;
  const l10nLabel = translate(l10nKey);
  return l10nLabel !== l10nKey ? l10nLabel : baseName;
}

export function getLocalizedMetricName(metric: { key: string, name: string }) {
  const bundleKey = `metric.${metric.key}.name`;
  const fromBundle = translate(bundleKey);
  return fromBundle !== bundleKey ? fromBundle : metric.name;
}

export function getLocalizedMetricDomain(domainName: string) {
  const bundleKey = `metric_domain.${domainName}`;
  const fromBundle = translate(bundleKey);
  return fromBundle !== bundleKey ? fromBundle : domainName;
}
