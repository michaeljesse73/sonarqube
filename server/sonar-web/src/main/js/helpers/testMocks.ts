/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { InjectedRouter } from 'react-router';
import { Store, createStore } from 'redux';
import { Location } from 'history';
import { ParsedAnalysis } from '../apps/projectActivity/utils';
import { Profile } from '../apps/quality-profiles/types';
import { DocumentationEntry } from '../apps/documentation/utils';

export function mockAlmApplication(overrides: Partial<T.AlmApplication> = {}): T.AlmApplication {
  return {
    backgroundColor: '#444444',
    iconPath: '/images/sonarcloud/github-white.svg',
    installationUrl: 'https://github.com/apps/greg-sonarcloud/installations/new',
    key: 'github',
    name: 'GitHub',
    ...overrides
  };
}

export function mockAlmOrganization(overrides: Partial<T.AlmOrganization> = {}): T.AlmOrganization {
  return {
    avatar: 'http://example.com/avatar',
    almUrl: 'https://github.com/foo',
    description: 'description-foo',
    key: 'foo',
    name: 'foo',
    personal: false,
    privateRepos: 0,
    publicRepos: 3,
    url: 'http://example.com/foo',
    ...overrides
  };
}

export function mockParsedAnalysis(overrides: Partial<ParsedAnalysis> = {}): ParsedAnalysis {
  return {
    date: new Date('2017-03-01T09:36:01+0100'),
    events: [],
    key: 'foo',
    projectVersion: '1.0',
    ...overrides
  };
}

export function mockAnalysisEvent(overrides: Partial<T.AnalysisEvent> = {}): T.AnalysisEvent {
  return {
    category: 'QUALITY_GATE',
    key: 'E11',
    description: 'Lorem ipsum dolor sit amet',
    name: 'Lorem ipsum',
    qualityGate: {
      status: 'ERROR',
      stillFailing: true,
      failing: [
        {
          key: 'foo',
          name: 'Foo',
          branch: 'master'
        },
        {
          key: 'bar',
          name: 'Bar',
          branch: 'feature/bar'
        }
      ]
    },
    ...overrides
  };
}

export function mockAppState(overrides: Partial<T.AppState> = {}): T.AppState {
  return {
    defaultOrganization: 'foo',
    edition: 'community',
    productionDatabase: true,
    qualifiers: ['TRK'],
    settings: {},
    version: '1.0',
    ...overrides
  };
}

export function mockComponent(overrides: Partial<T.Component> = {}): T.Component {
  return {
    breadcrumbs: [],
    key: 'my-project',
    name: 'MyProject',
    organization: 'foo',
    qualifier: 'TRK',
    qualityGate: { isDefault: true, key: '30', name: 'Sonar way' },
    qualityProfiles: [
      {
        deleted: false,
        key: 'my-qp',
        language: 'ts',
        name: 'Sonar way'
      }
    ],
    tags: [],
    ...overrides
  };
}

export function mockComponentMeasure(
  file = false,
  overrides: Partial<T.ComponentMeasure> = {}
): T.ComponentMeasure {
  if (file) {
    return {
      key: 'foo:src/index.tsx',
      name: 'index.tsx',
      qualifier: 'FIL',
      path: 'src/index.tsx',
      measures: [{ metric: 'bugs', value: '1', bestValue: false }],
      ...overrides
    };
  }
  return {
    key: 'foo',
    name: 'Foo',
    qualifier: 'TRK',
    measures: [{ metric: 'bugs', value: '12', bestValue: false }],
    ...overrides
  };
}

export function mockQualityGateStatusCondition(
  overrides: Partial<T.QualityGateStatusCondition> = {}
): T.QualityGateStatusCondition {
  return {
    actual: '10',
    error: '0',
    level: 'ERROR',
    metric: 'foo',
    op: 'GT',
    ...overrides
  };
}

export function mockCurrentUser(overrides: Partial<T.CurrentUser> = {}): T.CurrentUser {
  return {
    isLoggedIn: false,
    ...overrides
  };
}

export function mockLoggedInUser(overrides: Partial<T.LoggedInUser> = {}): T.LoggedInUser {
  return {
    groups: [],
    isLoggedIn: true,
    login: 'luke',
    name: 'Skywalker',
    scmAccounts: [],
    ...overrides
  };
}

export function mockEvent(overrides = {}) {
  return {
    target: { blur() {} },
    currentTarget: { blur() {} },
    preventDefault() {},
    stopPropagation() {},
    ...overrides
  } as any;
}

export function mockIssue(withLocations = false, overrides: Partial<T.Issue> = {}) {
  const issue: T.Issue = {
    actions: [],
    component: 'main.js',
    componentLongName: 'main.js',
    componentQualifier: 'FIL',
    componentUuid: 'foo1234',
    creationDate: '2017-03-01T09:36:01+0100',
    flows: [],
    fromHotspot: false,
    key: 'AVsae-CQS-9G3txfbFN2',
    line: 25,
    message: 'Reduce the number of conditional operators (4) used in the expression',
    organization: 'myorg',
    project: 'myproject',
    projectKey: 'foo',
    projectName: 'Foo',
    projectOrganization: 'org',
    rule: 'javascript:S1067',
    ruleName: 'foo',
    secondaryLocations: [],
    severity: 'MAJOR',
    status: 'OPEN',
    textRange: { startLine: 25, endLine: 26, startOffset: 0, endOffset: 15 },
    transitions: [],
    type: 'BUG'
  };

  function loc(): T.FlowLocation {
    return {
      component: 'main.js',
      textRange: { startLine: 1, startOffset: 1, endLine: 2, endOffset: 2 }
    };
  }

  if (withLocations) {
    issue.flows = [[loc(), loc(), loc()], [loc(), loc()]];
    issue.secondaryLocations = [loc(), loc()];
  }

  return {
    ...issue,
    ...overrides
  };
}

export function mockLocation(overrides: Partial<Location> = {}): Location {
  return {
    action: 'PUSH',
    hash: '',
    key: 'key',
    pathname: '/path',
    query: {},
    search: '',
    state: {},
    ...overrides
  };
}

export function mockMetric(overrides: Partial<T.Metric> = {}): T.Metric {
  return {
    id: 'coverage',
    key: 'coverage',
    name: 'Coverage',
    type: 'PERCENT',
    ...overrides
  };
}

export function mockMeasure(overrides: Partial<T.Measure> = {}): T.Measure {
  return {
    bestValue: true,
    metric: 'bugs',
    periods: [
      {
        bestValue: true,
        index: 1,
        value: '1.0'
      }
    ],
    value: '1.0',
    ...overrides
  };
}

export function mockOrganization(overrides: Partial<T.Organization> = {}): T.Organization {
  return { key: 'foo', name: 'Foo', ...overrides };
}

export function mockOrganizationWithAdminActions(
  overrides: Partial<T.Organization> = {},
  actionsOverrides: Partial<T.Organization['actions']> = {}
) {
  return mockOrganization({ actions: { admin: true, ...actionsOverrides }, ...overrides });
}

export function mockOrganizationWithAlm(
  overrides: Partial<T.Organization> = {},
  almOverrides: Partial<T.Organization['alm']> = {}
): T.Organization {
  return mockOrganization({
    alm: { key: 'github', membersSync: false, url: 'https://github.com/foo', ...almOverrides },
    ...overrides
  });
}

export function mockQualityGate(overrides: Partial<T.QualityGate> = {}): T.QualityGate {
  return {
    id: 1,
    name: 'qualitygate',
    ...overrides
  };
}

export function mockPullRequest(overrides: Partial<T.PullRequest> = {}): T.PullRequest {
  return {
    analysisDate: '2018-01-01',
    base: 'master',
    branch: 'feature/foo/bar',
    key: '1001',
    title: 'Foo Bar feature',
    ...overrides
  };
}

export function mockQualityProfile(overrides: Partial<Profile> = {}): Profile {
  return {
    activeDeprecatedRuleCount: 2,
    activeRuleCount: 10,
    childrenCount: 0,
    depth: 1,
    isBuiltIn: false,
    isDefault: false,
    isInherited: false,
    key: 'key',
    language: 'js',
    languageName: 'JavaScript',
    name: 'name',
    projectCount: 3,
    organization: 'foo',
    ...overrides
  };
}

export function mockQualityGateProjectStatus(
  overrides: Partial<T.QualityGateProjectStatus> = {}
): T.QualityGateProjectStatus {
  return {
    conditions: [
      {
        actualValue: '0',
        comparator: 'GT',
        errorThreshold: '1.0',
        metricKey: 'new_bugs',
        periodIndex: 1,
        status: 'OK'
      }
    ],
    ignoredConditions: false,
    status: 'OK',
    ...overrides
  };
}

export function mockRouter(overrides: { push?: Function; replace?: Function } = {}) {
  return {
    createHref: jest.fn(),
    createPath: jest.fn(),
    go: jest.fn(),
    goBack: jest.fn(),
    goForward: jest.fn(),
    isActive: jest.fn(),
    push: jest.fn(),
    replace: jest.fn(),
    setRouteLeaveHook: jest.fn(),
    ...overrides
  } as InjectedRouter;
}

export function mockRule(overrides: Partial<T.Rule> = {}): T.Rule {
  return {
    key: 'javascript:S1067',
    lang: 'js',
    langName: 'JavaScript',
    name: 'Use foo',
    severity: 'MAJOR',
    status: 'READY',
    sysTags: ['a', 'b'],
    tags: ['x'],
    type: 'CODE_SMELL',
    ...overrides
  } as T.Rule;
}

export function mockRuleDetails(overrides: Partial<T.RuleDetails> = {}): T.RuleDetails {
  return {
    key: 'squid:S1337',
    repo: 'squid',
    name: '".equals()" should not be used to test the values of "Atomic" classes',
    createdAt: '2014-12-16T17:26:54+0100',
    htmlDesc: '',
    mdDesc: '',
    severity: 'MAJOR',
    status: 'READY',
    isTemplate: false,
    tags: [],
    sysTags: ['multi-threading'],
    lang: 'java',
    langName: 'Java',
    params: [],
    defaultDebtRemFnType: 'CONSTANT_ISSUE',
    defaultDebtRemFnOffset: '5min',
    debtOverloaded: false,
    debtRemFnType: 'CONSTANT_ISSUE',
    debtRemFnOffset: '5min',
    defaultRemFnType: 'CONSTANT_ISSUE',
    defaultRemFnBaseEffort: '5min',
    remFnType: 'CONSTANT_ISSUE',
    remFnBaseEffort: '5min',
    remFnOverloaded: false,
    scope: 'MAIN',
    isExternal: false,
    type: 'BUG',
    ...overrides
  };
}

export function mockShortLivingBranch(
  overrides: Partial<T.ShortLivingBranch> = {}
): T.ShortLivingBranch {
  return {
    analysisDate: '2018-01-01',
    isMain: false,
    name: 'release-1.0',
    mergeBranch: 'master',
    type: 'SHORT',
    ...overrides
  };
}

export function mockLongLivingBranch(
  overrides: Partial<T.LongLivingBranch> = {}
): T.LongLivingBranch {
  return {
    analysisDate: '2018-01-01',
    isMain: false,
    name: 'master',
    type: 'LONG',
    ...overrides
  };
}

export function mockStore(state: any = {}, reducer = (state: any) => state): Store {
  return createStore(reducer, state);
}

export function mockDocumentationEntry(
  overrides: Partial<DocumentationEntry> = {}
): DocumentationEntry {
  return {
    content: 'Lorem ipsum dolor sit amet fredum',
    relativeName: 'Lorem',
    navTitle: undefined,
    text: 'Lorem ipsum dolor sit amet fredum',
    title: 'Lorem',
    url: '/lorem/ipsum',
    ...overrides
  };
}

export function mockMainBranch(overrides: Partial<T.MainBranch> = {}): T.MainBranch {
  return {
    analysisDate: '2018-01-01',
    isMain: true,
    name: 'master',
    ...overrides
  };
}

export function mockLanguage(overrides: Partial<T.Language> = {}): T.Language {
  return {
    key: 'css',
    name: 'CSS',
    ...overrides
  };
}
