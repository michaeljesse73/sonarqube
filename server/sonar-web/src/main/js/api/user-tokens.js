/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
// @flow
import { getJSON, postJSON, post } from '../helpers/request';
import throwGlobalError from '../app/utils/throwGlobalError';

/**
 * List tokens for given user login
 * @param {string} login
 * @returns {Promise}
 */
export function getTokens(login: string) {
  const url = '/api/user_tokens/search';
  const data = { login };
  return getJSON(url, data).then(r => r.userTokens);
}

/**
 * Generate a user token
 * @param {string} userLogin
 * @param {string} tokenName
 * @returns {Promise}
 */
export function generateToken(
  tokenName: string,
  userLogin?: string
): Promise<{ name: string, token: string }> {
  const url = '/api/user_tokens/generate';
  const data: { [string]: string } = { name: tokenName };
  if (userLogin) {
    data.login = userLogin;
  }
  return postJSON(url, data).catch(throwGlobalError);
}

/**
 * Revoke a user token
 * @param {string} userLogin
 * @param {string} tokenName
 * @returns {Promise}
 */
export function revokeToken(tokenName: string, userLogin?: string) {
  const url = '/api/user_tokens/revoke';
  const data: { [string]: string } = { name: tokenName };
  if (userLogin) {
    data.login = userLogin;
  }
  return post(url, data).catch(throwGlobalError);
}
