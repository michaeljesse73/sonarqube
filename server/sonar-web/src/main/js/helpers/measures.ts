/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { translate, translateWithParameters } from 'sonar-ui-common/helpers/l10n';
import { formatMeasure } from 'sonar-ui-common/helpers/measures';
import { isDefined } from 'sonar-ui-common/helpers/types';

/** Return a localized metric name */
export function localizeMetric(metricKey: string): string {
  return translate('metric', metricKey, 'name');
}

/** Return corresponding "short" for better display in UI */
export function getShortType(type: string): string {
  if (type === 'INT') {
    return 'SHORT_INT';
  } else if (type === 'WORK_DUR') {
    return 'SHORT_WORK_DUR';
  }
  return type;
}

export function enhanceMeasuresWithMetrics(
  measures: T.Measure[],
  metrics: T.Metric[]
): T.MeasureEnhanced[] {
  return measures
    .map(measure => {
      const metric = metrics.find(metric => metric.key === measure.metric);
      return metric && { ...measure, metric };
    })
    .filter(isDefined);
}

/** Get period value of a measure */
export function getPeriodValue(
  measure: T.Measure | T.MeasureEnhanced,
  periodIndex: number
): string | undefined {
  const { periods } = measure;
  const period = periods && periods.find(period => period.index === periodIndex);
  return period ? period.value : undefined;
}

export function isPeriodBestValue(
  measure: T.Measure | T.MeasureEnhanced,
  periodIndex: number
): boolean {
  const { periods } = measure;
  const period = periods && periods.find(period => period.index === periodIndex);
  return (period && period.bestValue) || false;
}

/** Check if metric is differential */
export function isDiffMetric(metricKey: string): boolean {
  return metricKey.indexOf('new_') === 0;
}

function getRatingGrid(): string {
  // workaround cyclic dependencies
  const getStore = require('../app/utils/getStore').default;
  const { getGlobalSettingValue } = require('../store/rootReducer');

  const store = getStore();
  const settingValue = getGlobalSettingValue(store.getState(), 'sonar.technicalDebt.ratingGrid');
  return settingValue ? settingValue.value : '';
}

let maintainabilityRatingGrid: number[];

function getMaintainabilityRatingGrid(): number[] {
  if (maintainabilityRatingGrid) {
    return maintainabilityRatingGrid;
  }

  const str = getRatingGrid();
  const numbers = str
    .split(',')
    .map(s => parseFloat(s))
    .filter(n => !isNaN(n));

  if (numbers.length === 4) {
    maintainabilityRatingGrid = numbers;
  } else {
    maintainabilityRatingGrid = [0, 0, 0, 0];
  }

  return maintainabilityRatingGrid;
}

function getMaintainabilityRatingTooltip(rating: number): string {
  const maintainabilityGrid = getMaintainabilityRatingGrid();
  const maintainabilityRatingThreshold = maintainabilityGrid[Math.floor(rating) - 2];

  if (rating < 2) {
    return translateWithParameters(
      'metric.sqale_rating.tooltip.A',
      formatMeasure(maintainabilityGrid[0] * 100, 'PERCENT')
    );
  }

  const ratingLetter = formatMeasure(rating, 'RATING');

  return translateWithParameters(
    'metric.sqale_rating.tooltip',
    ratingLetter,
    formatMeasure(maintainabilityRatingThreshold * 100, 'PERCENT')
  );
}

export function getRatingTooltip(metricKey: string, value: number | string): string {
  const ratingLetter = formatMeasure(value, 'RATING');

  const finalMetricKey = isDiffMetric(metricKey) ? metricKey.substr(4) : metricKey;

  return finalMetricKey === 'sqale_rating' || finalMetricKey === 'maintainability_rating'
    ? getMaintainabilityRatingTooltip(Number(value))
    : translate('metric', finalMetricKey, 'tooltip', ratingLetter);
}

export function getDisplayMetrics(metrics: T.Metric[]) {
  return metrics.filter(metric => !metric.hidden && !['DATA', 'DISTRIB'].includes(metric.type));
}

export function findMeasure(measures: T.Measure[], metric: string) {
  return measures.find(measure => measure.metric === metric);
}
