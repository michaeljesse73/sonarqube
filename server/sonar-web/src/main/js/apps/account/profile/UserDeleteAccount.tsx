/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { Button } from 'sonar-ui-common/components/controls/buttons';
import DeferredSpinner from 'sonar-ui-common/components/ui/DeferredSpinner';
import { translate } from 'sonar-ui-common/helpers/l10n';
import { getOrganizationsThatPreventDeletion } from '../../../api/organizations';
import { whenLoggedIn } from '../../../components/hoc/whenLoggedIn';
import { withUserOrganizations } from '../../../components/hoc/withUserOrganizations';
import UserDeleteAccountContent from './UserDeleteAccountContent';
import UserDeleteAccountModal from './UserDeleteAccountModal';

interface Props {
  user: T.LoggedInUser;
  userOrganizations: T.Organization[];
}

interface State {
  loading: boolean;
  organizationsToTransferOrDelete: T.Organization[];
  showModal: boolean;
}

export class UserDeleteAccount extends React.PureComponent<Props, State> {
  mounted = false;

  state: State = {
    loading: true,
    organizationsToTransferOrDelete: [],
    showModal: false
  };

  componentDidMount() {
    this.mounted = true;
    this.fetchOrganizationsThatPreventDeletion();
  }

  componentWillUnmount() {
    this.mounted = false;
  }

  fetchOrganizationsThatPreventDeletion = () => {
    getOrganizationsThatPreventDeletion().then(
      ({ organizations }) => {
        if (this.mounted) {
          this.setState({
            loading: false,
            organizationsToTransferOrDelete: organizations
          });
        }
      },
      () => {}
    );
  };

  toggleModal = () => {
    if (this.mounted) {
      this.setState(state => ({
        showModal: !state.showModal
      }));
    }
  };

  render() {
    const { user, userOrganizations } = this.props;
    const { organizationsToTransferOrDelete, loading, showModal } = this.state;

    const label = translate('my_profile.delete_account');

    return (
      <div>
        <h2 className="spacer-bottom">{label}</h2>

        <DeferredSpinner loading={loading} />

        {!loading && (
          <>
            <UserDeleteAccountContent
              className="list-styled no-padding big-spacer-top big-spacer-bottom"
              organizationsSafeToDelete={userOrganizations}
              organizationsToTransferOrDelete={organizationsToTransferOrDelete}
            />

            <Button
              className="button-red"
              disabled={organizationsToTransferOrDelete.length > 0}
              onClick={this.toggleModal}
              type="button">
              {translate('delete')}
            </Button>

            {showModal && (
              <UserDeleteAccountModal
                label={label}
                organizationsSafeToDelete={userOrganizations}
                organizationsToTransferOrDelete={organizationsToTransferOrDelete}
                toggleModal={this.toggleModal}
                user={user}
              />
            )}
          </>
        )}
      </div>
    );
  }
}

export default whenLoggedIn(withUserOrganizations(UserDeleteAccount));
