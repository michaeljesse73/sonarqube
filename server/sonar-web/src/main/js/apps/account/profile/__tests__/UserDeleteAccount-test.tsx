/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow } from 'enzyme';
import * as React from 'react';
import { click, waitAndUpdate } from 'sonar-ui-common/helpers/testUtils';
import { getOrganizationsThatPreventDeletion } from '../../../../api/organizations';
import { mockLoggedInUser, mockOrganization } from '../../../../helpers/testMocks';
import { UserDeleteAccount } from '../UserDeleteAccount';

jest.mock('../../../../api/organizations', () => ({
  getOrganizationsThatPreventDeletion: jest.fn().mockResolvedValue({ organizations: [] })
}));

beforeEach(() => {
  jest.clearAllMocks();
});

const organizationToTransferOrDelete = {
  key: 'luke-leia',
  name: 'Luke and Leia'
};

it('should render correctly', async () => {
  const wrapper = shallowRender();
  expect(wrapper).toMatchSnapshot();

  await waitAndUpdate(wrapper);
  expect(wrapper).toMatchSnapshot();

  click(wrapper.find('Button'));
  expect(wrapper).toMatchSnapshot();
});

it('should get some organizations', async () => {
  (getOrganizationsThatPreventDeletion as jest.Mock).mockResolvedValue({
    organizations: [organizationToTransferOrDelete]
  });

  const wrapper = shallowRender();

  await waitAndUpdate(wrapper);

  expect(wrapper.state('loading')).toBeFalsy();
  expect(wrapper.state('organizationsToTransferOrDelete')).toEqual([
    organizationToTransferOrDelete
  ]);
  expect(getOrganizationsThatPreventDeletion).toBeCalled();
  expect(wrapper.find('Button').prop('disabled')).toBe(true);
});

it('should toggle modal', () => {
  const wrapper = shallowRender();
  wrapper.setState({ loading: false });
  expect(wrapper.find('Connect(withRouter(UserDeleteAccountModal))').exists()).toBe(false);
  click(wrapper.find('Button'));
  expect(wrapper.find('Connect(withRouter(UserDeleteAccountModal))').exists()).toBe(true);
});

function shallowRender(props: Partial<UserDeleteAccount['props']> = {}) {
  const user = mockLoggedInUser({ externalIdentity: 'luke' });

  const userOrganizations = [
    mockOrganization({ key: 'luke-leia', name: 'Luke and Leia' }),
    mockOrganization({ key: 'luke', name: 'Luke Skywalker' })
  ];

  return shallow<UserDeleteAccount>(
    <UserDeleteAccount user={user} userOrganizations={userOrganizations} {...props} />
  );
}
