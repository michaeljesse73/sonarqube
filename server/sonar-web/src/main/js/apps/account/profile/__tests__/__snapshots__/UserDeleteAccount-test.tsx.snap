// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div>
  <h2
    className="spacer-bottom"
  >
    my_profile.delete_account
  </h2>
  <DeferredSpinner
    loading={true}
    timeout={100}
  />
</div>
`;

exports[`should render correctly 2`] = `
<div>
  <h2
    className="spacer-bottom"
  >
    my_profile.delete_account
  </h2>
  <DeferredSpinner
    loading={false}
    timeout={100}
  />
  <UserDeleteAccountContent
    className="list-styled no-padding big-spacer-top big-spacer-bottom"
    organizationsSafeToDelete={
      Array [
        Object {
          "key": "luke-leia",
          "name": "Luke and Leia",
        },
        Object {
          "key": "luke",
          "name": "Luke Skywalker",
        },
      ]
    }
    organizationsToTransferOrDelete={Array []}
  />
  <Button
    className="button-red"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    delete
  </Button>
</div>
`;

exports[`should render correctly 3`] = `
<div>
  <h2
    className="spacer-bottom"
  >
    my_profile.delete_account
  </h2>
  <DeferredSpinner
    loading={false}
    timeout={100}
  />
  <UserDeleteAccountContent
    className="list-styled no-padding big-spacer-top big-spacer-bottom"
    organizationsSafeToDelete={
      Array [
        Object {
          "key": "luke-leia",
          "name": "Luke and Leia",
        },
        Object {
          "key": "luke",
          "name": "Luke Skywalker",
        },
      ]
    }
    organizationsToTransferOrDelete={Array []}
  />
  <Button
    className="button-red"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    delete
  </Button>
  <Connect(withRouter(UserDeleteAccountModal))
    label="my_profile.delete_account"
    organizationsSafeToDelete={
      Array [
        Object {
          "key": "luke-leia",
          "name": "Luke and Leia",
        },
        Object {
          "key": "luke",
          "name": "Luke Skywalker",
        },
      ]
    }
    organizationsToTransferOrDelete={Array []}
    toggleModal={[Function]}
    user={
      Object {
        "externalIdentity": "luke",
        "groups": Array [],
        "isLoggedIn": true,
        "login": "luke",
        "name": "Skywalker",
        "scmAccounts": Array [],
      }
    }
  />
</div>
`;
